accepted/2024/NuGet-Audit-for-PackageDownload.md

```diff
@@ -0,0 +1,123 @@
+
+```xml
+<ItemGroup>
+    <PackageDownload Include="NuGet.Protocol" Version="[5.11.2]" />
+</ItemGroup>
+```
+
+If "NuGet.Protocol" is found to have a vulnerability, NuGet Audit will issue a warning to alert you so you can update to a safer version.
+
+##### Warning NU1906
+
+> warning NU1906: Package 'NuGet.Protocol' 5.11.2 has a known moderate severity vulnerability, https://github.com/advisories/GHSA-g3q9-xf95-8hp5
+
+
+On the other hand imagine the SDK has the package `System.Text.Json`. The SDK would download this package as follows
+
+```xml
+<ItemGroup>
+    <PackageDownload Include="System.Text.Json" Version="[1.0.0]" IsImplicitlyDefined="true"/>
+</ItemGroup>
+```
+
+Since this was implicitly added, no warnings are received by the user.
+
+#### Functional Design
+
+Here’s a pseudocode version of how this exclusion works:
+
+```csharp
+foreach (package in PackageDownloads):
+    if (package.HasAttribute("IsImplicitlyDefined") && package.IsImplicitlyDefined == true):
+        continue  // Skip auditing, as it's an SDK-managed package
+    else:
+        checkForVulnerabilities(package)
+```
+
+Now this ensures users are provided with NuGet Audit warnings that are not confusing and actionable.
+
+### Technical explanation
+
+<!-- Explain the proposal in sufficient detail with implementation details, interaction models, and clarification of corner cases. -->
+
+The following technical components will support the proposal's functionality:
+
+1. **New Attribute for SDK-Specific PackageDownloads:** A new attribute, named `IsImplicitlyDefined="true"`, will be added to <PackageDownload> entries within the SDK.
+ This will also be done in project-system to make sure VS nominations are working in a similar manner. 
+ Add the property [here](https://github.com/dotnet/project-system/blob/9f35656ad68aa1352d7b6b0fd01784f7aefe1005/src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/CollectedPackageDownload.xaml#L3) for project-system.
+
+1. **NuGet Audit Update to Handle Attribute:** NuGet Audit will be modified to recognize the new attribute and exclude SDK-flagged PackageDownloads from security alerts. 
+This update ensures that only relevant <PackageDownload> packages, added directly by the user, are scanned for vulnerabilities, allowing users to remain informed without SDK-related alerts.
+
+
+## Drawbacks
+
+<!-- Why should we not do this? -->
+
+## Rationale and alternatives
+
+<!-- Why is this the best design compared to other designs? -->
+<!-- What other designs have been considered and why weren't they chosen? -->
+<!-- What is the impact of not doing this? -->
+
+## Prior Art
+
+<!-- What prior art, both good and bad are related to this proposal? -->
+<!-- Do other features exist in other ecosystems and what experience have their community had? -->
+<!-- What lessons from other communities can we learn from? -->
+<!-- Are there any resources that are relevant to this proposal? -->
+
+## Unresolved Questions
```

Comment:
I think we need to update the SDK to include the new attribute on its PackageDownloads in the same release that we add this NuGet feature.


Question: What happens when restoring from Visual Studio but using an older SDK that doesn't have those metadata values?  Do we need some sort of lightup so NuGet only does this if the SDK has the right metadata?

Reply:
Good catch. Wouldn't want to flag user provided package downloads if they simply came from an older SDK. Should we try to get the metadata into 9.0.2xx just so there is some coverage there of the mismatch?